vfs/cache_local.go

```diff
@@ -0,0 +1,407 @@
+package vfs
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/cozy/cozy-stack/couchdb"
+	"github.com/cozy/cozy-stack/couchdb/mango"
+	"github.com/cozy/cozy-stack/lru"
+	"golang.org/x/sync/errgroup"
+)
+
+// LocalCache implements the VFS Cache interface and should be used
+// for mono-stack usage, where only one cozy-stack is accessing to the
+// VFS at a time.
+//
+// Internally it provides some optimisations to cache file attributes
+// and avoid having multiple useless RTTs with CouchDB.
+type LocalCache struct {
+	mud  sync.RWMutex       // mutex for directories data-structures
+	lrud *lru.Cache         // lru cache for directories
+	pthd map[string]*string // path directory to id map
+
+	muf  sync.RWMutex       // mutex for files data-structures
+	lruf *lru.Cache         // lru cache for files
+	pthf map[string]*string // (folderID, name) file pair to id map
+}
+
+// NewLocalCache creates an new LocalCache. The maxEntries parameter
+// is used to specified the cache size: how many files and directories
+// elements are kept in-memory
+func NewLocalCache(maxEntries int) *LocalCache {
+	lc := new(LocalCache)
+	lc.init(maxEntries)
+	return lc
+}
+
+func (lc *LocalCache) init(maxEntries int) {
+	dirEviction := func(key string, value interface{}) {
+		if doc, ok := value.(*DirDoc); ok {
+			delete(lc.pthd, doc.Fullpath)
+		}
+	}
+
+	fileEviction := func(key string, value interface{}) {
+		if doc, ok := value.(*FileDoc); ok {
+			delete(lc.pthf, genFilePathID(doc.FolderID, doc.Name))
+		}
+	}
+
+	lc.pthd = make(map[string]*string)
+	lc.pthf = make(map[string]*string)
+	lc.lrud = &lru.Cache{MaxEntries: maxEntries, OnEvicted: dirEviction}
+	lc.lruf = &lru.Cache{MaxEntries: maxEntries, OnEvicted: fileEviction}
+}
+
+// CreateDir is be used to persist a directory document
+func (lc *LocalCache) CreateDir(c Context, doc *DirDoc) error {
+	var err error
+	if err = doc.calcPath(c); err != nil {
+		return err
+	}
+	err = couchdb.CreateDoc(c, doc)
+	if err != nil {
+		return err
+	}
+	lc.touchDir(doc)
+	return nil
+}
+
+// UpdateDir is used to update a persisted directory document
+func (lc *LocalCache) UpdateDir(c Context, olddoc, newdoc *DirDoc) error {
+	oldpath, err := olddoc.Path(c)
+	if err != nil {
+		return err
+	}
+
+	if err = lc.updateDirDoc(c, newdoc); err != nil {
+		return err
+	}
+
+	newpath, err := newdoc.Path(c)
+	if err != nil {
+		return err
+	}
+
+	if oldpath == newpath {
+		return nil
+	}
+
+	var children []*DirDoc
```

Comment:
Indeed I'll have to fix that.

I'll probably use a radix tree instead of a map to store directories by path. I made [this lib](https://github.com/jinroh/radix) few weeks ago, time to use it.